### PR TITLE
[5.1] #9970 $request->get() now includes json content

### PR DESCRIPTION
**Pull request by samlev**
`$request->get()` wouldn't include json body content but `$request->input()` would (because it internally uses `getInputSource()` which actually responds to `isJson()`).

The result is that `$request->all()` would return content that is passed through the request body with the header `CONTENT_TYPE=application/json`, but won't return anything that is passed as a parameter, while `$request->get()` was unaware of any data passed through the request body.

This pull request modifies the `get` method to allow access to json content (if it exists), which will make the `Request` object behave correctly when `CONTENT_TYPE=application/json` is set.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Tests?

**Comment by kidshenlong** [comment]
I didn't think the issue was with the actual request class! I thought it was just the fact that Laravel's HTTP test methods didn't pass json as the right parameter in the call method.

**Comment by samlev** [comment]
@GrahamCampbell added a test for this.

File: tests/Http/HttpRequestTest.php [hunk 1]
```diff
@@ -537,4 +537,22 @@ public function testCreateFromBase()
...
+
+    public function testJsonContent()
+    {
+        $body = [
+            'foo' => 'bar',
+            'baz' => ['qux'],
+        ];
+
+        $server = [
+            'CONTENT_TYPE' => 'application/json',
```
Comment (GrahamCampbell) [inline tests/Http/HttpRequestTest.php]: this is wrong isn't it?

Comment (GrahamCampbell) [inline tests/Http/HttpRequestTest.php]: you need the HTTP prefix

Comment (samlev) [inline tests/Http/HttpRequestTest.php]: Nope, CONTENT_TYPE is the appropriate header.

From elsewhere in the REQUEST object:

```php
/**
 * Determine if the request is sending JSON.
 *
 * @return bool
 */
public function isJson()
{
    return Str::contains($this->header('CONTENT_TYPE'), '/json');
}
```

Comment (GrahamCampbell) [inline tests/Http/HttpRequestTest.php]: no, that's different...

Comment (GrahamCampbell) [inline tests/Http/HttpRequestTest.php]: when you put the headers in, you need the HTTP prefix

Comment (samlev) [inline tests/Http/HttpRequestTest.php]: Well technically, both will work, but the Symfony request will translate `HTTP_CONTENT_TYPE` to `CONTENT_TYPE`:

```php
/**
 * Creates a new request with values from PHP's super globals.
 *
 * @return Request A new request
 *
 * @api
 */
public static function createFromGlobals()
{
    // With the php's bug #66606, the php's built-in web server
    // stores the Content-Type and Content-Length header values in
    // HTTP_CONTENT_TYPE and HTTP_CONTENT_LENGTH fields.
    $server = $_SERVER;
    if ('cli-server' === php_sapi_name()) {
        if (array_key_exists('HTTP_CONTENT_LENGTH', $_SERVER)) {
            $server['CONTENT_LENGTH'] = $_SERVER['HTTP_CONTENT_LENGTH'];
        }
        if (array_key_exists('HTTP_CONTENT_TYPE', $_SERVER)) {
            $server['CONTENT_TYPE'] = $_SERVER['HTTP_CONTENT_TYPE'];
        }
    }
    $request = self::createRequestFromFactory($_GET, $_POST, array(), $_COOKIE, $_FILES, $server);
    if (0 === strpos($request->headers->get('CONTENT_TYPE'), 'application/x-www-form-urlencoded')
        && in_array(strtoupper($request->server->get('REQUEST_METHOD', 'GET')), array('PUT', 'DELETE', 'PATCH'))
    ) {
        parse_str($request->getContent(), $data);
        $request->request = new ParameterBag($data);
    }
    return $request;
}
```

In particular:

```php
if (array_key_exists('HTTP_CONTENT_TYPE', $_SERVER)) {
    $server['CONTENT_TYPE'] = $_SERVER['HTTP_CONTENT_TYPE'];
}
```

Comment (GrahamCampbell) [inline tests/Http/HttpRequestTest.php]: oh wait, I see - symfony deals that that

Comment (GrahamCampbell) [inline tests/Http/HttpRequestTest.php]: https://github.com/symfony/HttpFoundation/blob/2.7/Request.php#L287

Comment (GrahamCampbell) [inline tests/Http/HttpRequestTest.php]: > Well technically, both will work

No they won't. Only the one without HTTP will work.

Comment (GrahamCampbell) [inline tests/Http/HttpRequestTest.php]: Since we're only calling create.

Comment (GrahamCampbell) [inline tests/Http/HttpRequestTest.php]: So ignore my original request...

Comment (samlev) [inline tests/Http/HttpRequestTest.php]: The ServerBag does the translation too:

https://github.com/symfony/HttpFoundation/blob/2.7/ServerBag.php#L31

Comment (GrahamCampbell) [inline tests/Http/HttpRequestTest.php]: Ah right. Bit strange of them to duplicate that logic. Maybe we should send a PR to remove that?

Comment (samlev) [inline tests/Http/HttpRequestTest.php]: There's a lot of "odd" in symfony, which I'd rather not get too far into (other than making this work for Laravel). Also the 'yoda conditions' littered through their code really bug me.

**Comment by samlev** [comment]
Do the CI tests need a bump?

**Comment by taylorotwell** [comment]
Would we really want to customize Symfony's `get` method? It's not really documented anywhere in our docs is it? Why not just use `input`? as we have documented?

**Comment by samlev** [comment]
`get()` is more consistent with how most of the other interfaces are used (seeing as they're all wrapped around the same kind of Symfony ParameterBag).

This extension still gets the benefits of the Symfony `get()` but effectively adds another checked method in at the end.

**Comment by samlev** [comment]
Squashed to 2 commits

File: src/Illuminate/Http/Request.php [hunk 2]
```diff
@@ -879,6 +882,37 @@ public function offsetUnset($offset)
...
+     *
+     *  * slow
+     *  * prefer to get from a "named" source
+     *
+     * It is better to explicitly get request parameters from the appropriate
+     * public property instead (query, attributes, request, json).
+     *
+     * @param  string  $key
+     * @param  mixed   $default
+     * @param  bool    $deep
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Request.php]: don't align

File: src/Illuminate/Http/Request.php [hunk 3]
```diff
@@ -879,6 +882,37 @@ public function offsetUnset($offset)
     }
 
     /**
+     * Gets a "parameter" value.
+     *
+     * Order of precedence: GET, PATH, POST, JSON
+     *
+     * Avoid using this method in controllers:
+     *
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Request.php]: remove the extra space please

Comment (GrahamCampbell) [inline src/Illuminate/Http/Request.php]: infact, just ditch the list and just write a sentence.

File: src/Illuminate/Http/Request.php [hunk 4]
```diff
@@ -728,6 +728,9 @@ public static function createFromBase(SymfonyRequest $request)
             $request->cookies->all(), $request->files->all(), $request->server->all()
         );
 
+        // ensure that json is empty
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Request.php]: this comment is totally pointless

**Comment by taylorotwell** [comment]
I still don't really want to get into overriding a common Symfony method. It should be fixed in Symfony if it's a bug. Otherwise, we can use $request->input() to access JSON data.

**Comment by samlev** [comment]
It's not really a bug in Symfony because Symfony doesn't know/care about JSON body content. I would consider this a bug in Laravel because it has introduced the JSON content option (And associated functions), but done so inconsistently.

**Pull request closed without merging**